src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/storage"
+	xts "github.com/m3db/m3/src/query/ts"
+)
+
+type encodedStepIterUnconsolidated struct {
+	mu          sync.RWMutex
+	lastBlock   bool
+	exhausted   bool
+	started     bool
+	err         error
+	meta        block.Metadata
+	validIters  []bool
+	seriesMeta  []block.SeriesMeta
+	seriesIters []encoding.SeriesIterator
+}
+
+func (it *encodedStepIterUnconsolidated) Current() (
+	block.UnconsolidatedStep,
+	error,
+) {
+	it.mu.RLock()
+	if !it.started || it.exhausted {
+		it.mu.RUnlock()
```

Comment:
Deferring comes with a bit of a hit and this will get called pretty often. Probably a bit premature optimization, but there shouldn't be a lot of early returns, especially after the `Err` method is created on the iterator

Reply:
*Locking* comes with a bit of a hit :-p. Let's have that discussion about the locking--I'm still not sure what the use case is for the concurrent iterator.